Rewrite and extend computer graphics notes to follow a more conventional style

levyry opened the pull request:
This PR aims to rewrite the current computer graphics notes to follow a more conventional, textbook like style, similar to the official BSz1 and BSz2 notes, while also adding some quality of life features, such as:

- a preamble
- improved quizzes
- extra information and trivia
- references to the official youtube channel

### Guidelines 

- Use more admonitions
- Use inline mode less, or at least use `\displaymode`
- Use less lists and enumerations, use paragraphs instead

### Progress

- [x] rewrite notes
  - [x] all chapters done (12/12)
  - [x] improved images
  - [x] improved quizzes
- [x] add preamble
  - [x] add youtube playlist link
- [x] add common math terms page
- ~~add custom admonitions~~
  - ~~for definitions~~
  - ~~for extra knowledge~~

Kristof-me committed:
Made workflow manually runnable

levyry committed:
Rewriting 1.md to a different style

Kristof-me commented:
Cool idea! 
The guidelines are great but should be executed carefully. For example paragraphs are great, but enumerations may help with memorizing the content.

In `1.md` the admonitions are a bit too much for me personally. But these might actually help some people, so feel free to keep it as it is.

Gilgames32 commented on docs/notes/sem4/computer_graphics/1.md:
```diff
@@ -1,296 +1,409 @@
-- Ezeknek az axiómáknak a megváltoztatása különböző eredményekhez vezethet. Például a háromszög szögeinek összege mindig:
+A geometriák különböző axiómákra épülnek. Például az Euklidészi síkgeometriában az egyik legfontosabb, hogy egy egyenesre egy külső pontból legfeljebb egy olyan egyenes húzható, ami nem metszi (ez a párhuzamossági posztulátum)
+
+Ezeknek az axiómáknak a megváltoztatása különböző eredményekhez vezethet. Például a háromszög belső szögeinek összege
```
ilyen helyekre tegyél egy extra entert, mert egybe fogja rakni a bekezdéssel

Gilgames32 commented on docs/notes/sem4/computer_graphics/1.md:
```diff
@@ -1,296 +1,409 @@
+  - $T = (\alpha + \beta + \gamma - \pi) / \kappa$
+
+- A kör kerülete
+    ![A kör kerülete](./img/1_kerulet.png)
```
ide szintúgy, mert egy sorba fog kerülni a kép és a szöveg, nem néz ki jól

Gilgames32 commented on docs/notes/sem4/computer_graphics/1.md:
```diff
@@ -1,296 +1,409 @@
+
+<!-- Ez egy eléggé részletes példa egy olyan fogalomról, amit valószínűleg a legtöbb hallgató már ismer. Nem muszáj meghagyni ennyire részletesre, de úgy gondoltam, hogy akinek szüksége van magyarázatra, és direkt rákattint a kinyíló emlékeztetőre, annak nagy eséllyel szüksége is van egy ennyire részletes magyarázatra. -->
+??? info Emlékeztető (asszociativitás fogalma)
+    Egy (kétváltozós) műveletre (legyen most $*$) akkor mondjuk, hogy asszociatív, ha nem számít az elvégzésének sorrendje, azaz $$(u * v) * w = u * (v * w) = u * v * w.$$ Ha egy művelet _nem_ asszociatív, akkor az azt jelenti, hogy $(u * v) * w \neq u * (v * w)$. Egy példa egy nem asszociatív műveletre az osztás: $$\frac{1}{6} = \frac{\frac{1}{2}}{3} = \underbrace{(1/2)/3 \neq 1/(2/3)}_{
```
itt szintúgy

ha inline raksz nem inline matekos dolgot, akkor breakelni fog, mert inline rakja, de körülveszi `\( \)`-ekkel

Gilgames32 commented on docs/notes/sem4/computer_graphics/1.md:
```diff
@@ -1,296 +1,409 @@
+
+!!! question De mégis miért jobb ez?
+    Azért, mert ez magától kezeli a végtelen pontokat a Descartes koordinátákkal szemben:
+    $$
```
ide is kell +1 extra üres sor mert breakel

Gilgames32 commented on docs/notes/sem4/computer_graphics/1.md:
```diff
@@ -1,296 +1,409 @@
-> 8-9\. Mi igaz a geometriákra
+!!! question 8-9\. Mi igaz a geometriákra
 
 |   | Gömbi | Hiperbolikus |
```
ezt nem akartad szintén dropdown megcsinálni? nagyon tetszik hogy a megoldás alapból nincs mutatva, de ez így kicsit inkonzisztens

ha csak azért van így mert táblázat és a admonition miatt breakelne akkor megértem de így fura

Gilgames32 reviewed (dismissed):
Jó az ötlet és tetszik az irány, viszont a honlapon breakel a formázás. A vscodeos extensionnel lehet jól néz ki, de az másik engine. Kérlek buildeld le és nézd meg, hogy élesben is jól néz-e ki!

Gilgames32 commented on docs/notes/sem4/computer_graphics/1.md:
```diff
@@ -1,296 +1,409 @@
+
+Ezt valamennyire láttuk, a fontos különbség, hogy a $w$ a távolság jelölést is segíti nekünk:
+
+<!-- Nem tudom, hogy direkt használunk-e ilyen négyzetes sorvektorokat, ezért ebben a fejezetben direkt oszloposan írtam fel őket, hogy meglássuk, hogy milyen, és döntsük el, hogy melyiket preferálnánk. -->
```
ezzel óvatosan, később még lesz jelentősége hogy row vagy column major mátrixokat használunk-e, most így hirtelen viszont nem tudom ez vezetne-e inkonzisztenciához, ezért meghagynám ahogy volt

levy04 commented:
> For example paragraphs are great, but enumerations may help with memorizing the content.

This is something I have considered, and in my opinion, the main studying notes should be kept at least a bit separate from the "I already studied the material and I just want a quick refresher to cram stuff" type of notes. I don't have a really good solution to this, one idea I had was to include a "Review" section at the end, which would be a bullet point list of the most important topics of that chapter. I am open to other ideas though. 

> In `1.md` the admonitions are a bit too much for me personally. But these might actually help some people, so feel free to keep it as it is.

I might tone them down a bit, this first commit was more about putting stuff out there and seeing what works and what doesn't. In the future I'll try to keep admonitions to mainly definitions and the like.

levyry committed:
(docs 1/12): Finish first lecture notes.

levyry committed:
(docs 2/12): Finish second chapter notes.

restructure img paths

(docs): 2.md

(docs): fix 1.md

levyry committed:
(docs 2/12): Fix ndc quiz

levyry committed:
(docs 2/12): typos

levyry committed:
resolve merge conflict

initial notes

felületek jegyzet

levyry committed:
docs: 3.md done

docs: 3.md done

levyry committed:
docs: 4.md done

docs: 4.md fix

levyry committed:
add notes for several chapters

8.md notes

8.md 2

9.md notes

9.md notes wow such more notes

more notes

notes for 6.md

levyry committed:
finish 10.md

levyry committed:
refactor quizes

levyry committed:
tidy up 1st and 3rd chapters

tidy up 3.md

levyry committed:
add quiz question

levyry committed:
finish chapter 6

levyry committed:
remove settings.json

levyry committed:
remove \newlines in math mode

levyry committed:
finish chapter 11

chapter 11 halfway done

chapter 11 three-quarters done

chapter 11 done

levyry committed:
reformat quizes and math blocks, fix spelling errors

levyry commented:
After this rewrite I was planning on opening a second PR in which I'd implement general QoL changes to the notes, such as:
- adding a preamble chapter, with general information about the course (mainly related to studying it, not administrative info), with a place for anyone to add their opinions in PR's
- adding a link to the official computer graphics video from the lecturer at the beginning of each chapter, so it's available to anyone easily
- adding higher quality pictures in some chapters
- adding a TL;DR at the end of each chapter before the quiz, which would contain the most important formulas and ideas from the chapter
- with permission, adding my own [grafika-hf-template](https://github.com/levy04/grafika-hf-template) to the preamble, which I plan to maintain if the framework were to change again, and make platform independent
- adding a "common math terms" chapter, in which I would provide definitions, explanations and examples for certain terminology used in the material, such as: associative, commutative, algebra, orthogonal, normalized, unit vector, position vector, etc.

As these changes aren't strictly relevant to this PR, I won't include them here, but I would still appreciate some early feedback on these ideas, so they will have already been discussed a bit before opening the second PR.

levyry committed:
chapter 7 done

levyry committed:
chapter 7 done

levyry committed:
chapter 11 hotfix

levyry committed:
chapter 11 hotfix

levyry committed:
finish chapter 8

chapter 8 halfway done

levyry committed:
finish chapter 8

chapter 8 halfway done

levyry committed:
fix minor grammar errors in 7.md

levyry committed:
fix minor grammar errors in 7.md

levyry committed:
format quiz of chapter 12

levyry committed:
format quiz of chapter 12

levyry committed:
finish chapter 5

write up chapter 5 example and MVP transformation

rename chapter 5 chapter names

write up chapter 5 vágás and viewport, touch up on quiz

(fix): consistent notation for NDC square

finish chapter 5

levyry committed:
finish chapter 5

write up chapter 5 example and MVP transformation

rename chapter 5 chapter names

write up chapter 5 vágás and viewport, touch up on quiz

(fix): consistent notation for NDC square

finish chapter 5

levyry committed:
fix chapter 4 affin transformation

levyry committed:
fix chapter 4 affin transformation

levyry committed:
reformat chapter 10 quiz

levyry committed:
reformat chapter 10 quiz

levyry committed:
Merge branch 'main' into grafika-rewrite

levyry committed:
Merge branch 'main' into grafika-rewrite

levyry committed:
(tidy): Add and reword notes to a lot of chapters

This commit aims to clarify and add "NOTE:" blocks for future maintainers
to have more context when changing certain chapters.

It also fixes some math block rendering issues introduced by the previous
merge conflict fix commit.

It also rewords some of the quizzes and material, but only in minor amounts.

levyry committed:
(tidy): Add and reword notes to a lot of chapters

This commit aims to clarify and add "NOTE:" blocks for future maintainers
to have more context when changing certain chapters.

It also fixes some math block rendering issues introduced by the previous
merge conflict fix commit.

It also rewords some of the quizzes and material, but only in minor amounts.

levyry committed:
(tidy): use $\ast$ instead of $\*$

levyry committed:
(tidy): use $\ast$ instead of $\*$

levyry committed:
refactor quizzes

refactor quiz of chapter 1

refactor quiz of chapter 6

refactor quiz of chapter 9

refactor quiz of chapter 12

levyry committed:
refactor quizzes

refactor quiz of chapter 1

refactor quiz of chapter 6

refactor quiz of chapter 9

refactor quiz of chapter 12

levyry committed:
use automatic parenthesis size detection

levyry committed:
use automatic parenthesis size detection

levyry committed:
remove manual chapter numberings from 1.md

levyry committed:
remove manual chapter numberings from 1.md

levyry committed:
fix several quizzes

levyry committed:
fix several quizzes

levyry committed:
reword, correct several paragraphs, add quotes

This commit adds additional context to some parts of the written notes.

It also adds some quotes.

levyry committed:
reword, correct several paragraphs, add quotes

This commit adds additional context to some parts of the written notes.

It also adds some quotes.

levyry committed:
(fix): latex and code blocks

This commit fixes some inconsistencies between certain LaTeX and code
blocks.

levyry committed:
(fix): latex and code blocks

This commit fixes some inconsistencies between certain LaTeX and code
blocks.

levyry committed:
add addition NOTE blocks

levyry committed:
add addition NOTE blocks

levyry committed:
rephrase to remove self

levyry committed:
rephrase to remove self

levyry committed:
start work on chapter 12

levyry committed:
start work on chapter 12

levyry committed:
revert accidental changes to nonrelevant files

levyry committed:
revert accidental changes to nonrelevant files

levyry committed:
work on chapter 12

levyry committed:
work on chapter 12

Gilgames32 commented:
gg

levyry committed:
refactor: iron out chapter 1, fix image scaling issues

levyry committed:
refactor: iron out chapter 1, fix image scaling issues

levyry committed:
feat: continue work on chapter 12

levyry committed:
feat: continue work on chapter 12

levyry committed:
refactor: consistant image sizes, fix formatting mistakes in all chapters

levyry committed:
refactor: consistant image sizes, fix formatting mistakes in all chapters

levyry committed:
fix: allow linenumbered codeblocks

levyry committed:
fix: allow linenumbered codeblocks

levyry committed:
feat: add initial index page

levyry committed:
feat: add initial index page

levyry committed:
feat: finish chapter 12

levyry committed:
feat: finish chapter 12

levyry committed:
feat: add initial math appendix

levyry committed:
feat: add initial math appendix

levyry committed:
feat: expand math appendix, fix: prev chapters

levyry committed:
feat: expand math appendix, fix: prev chapters

levyry committed:
Revert accidental changes to Bsz1 notes

Find and replace to brrrr

levyry committed:
Revert accidental changes to Bsz1 notes

Find and replace to brrrr

levyry committed:
update maintainer information

levyry committed:
update maintainer information

levyry committed:
begin work of chapter 9

levyry committed:
begin work of chapter 9

levyry committed:
unify grammar of "koordináta-rendszer"

levyry committed:
unify grammar of "koordináta-rendszer"

levyry committed:
add note about matrixes and update maintainer info

levyry committed:
add note about matrixes and update maintainer info

levyry committed:
finish chapter 9

levyry committed:
finish chapter 9

levyry committed:
fix errors related to matrix representation

levyry committed:
fix errors related to matrix representation

levyry committed:
fix multiplication sign in ch8

levyry committed:
fix multiplication sign in ch8

Gilgames32 commented on docs/notes/sem4/computer_graphics/appendix/math_appendix.md:
````diff
@@ -0,0 +1,264 @@
+//...
+```
+
+Mivel az összeadás asszociatív, a program feldarabolhatja az összeadást. Például az `(1 + 2 + 3) + (4 + 5 + 6)` ugyanazt az eredményt adja, mint az `1 + (2 + 3) + 4 + (5 + 6)`. Így tehát akár több szálon is össze tudjuk adni a listának egy-egy részlistáját, amiket a végén szintén összeadunk. Ennél az elemi példánál persze a `for` ciklus is megfelelő, viszont bonyolultabb számítások esetén az, hog ...
````
> az, hogy lényegében "ingyen" kaptuk meg a párhuzamosságot az nagyon kényelmes

nem szép fogalmazás + vesszőhiba

bonyolultabb számítások esetén (felettébb) egyszerű/kényelmes a párhuzamosság megvalósítása. (vagy valami hasonló idk, nem az "ingyen" a baj inkább ez az "az [...] az" nem szép)

Gilgames32 commented on docs/notes/sem4/computer_graphics/appendix/math_appendix.md:
```diff
@@ -0,0 +1,264 @@
+
+Mivel az összeadás asszociatív, a program feldarabolhatja az összeadást. Például az `(1 + 2 + 3) + (4 + 5 + 6)` ugyanazt az eredményt adja, mint az `1 + (2 + 3) + 4 + (5 + 6)`. Így tehát akár több szálon is össze tudjuk adni a listának egy-egy részlistáját, amiket a végén szintén összeadunk. Ennél az elemi példánál persze a `for` ciklus is megfelelő, viszont bonyolultabb számítások esetén az, hog ...
+
+Ha elindulunk ezen az úton, akkor eljuthatunk a _monádok_ fogalmáig, amiket már minden programozó használt életében, és szerette őket, akkor is, ha pontosan nem tudta, hogy azt használta. Ha használtál már Javascriptben `Promise`-t, vagy Kotlinban `?.` operátort, vagy ha [hívtál már meg több függvényt egymás után](https://en.wikipedia.org/wiki/Function_composition_(computer_science)#Composing_fun ...
```
hogy *azokat* használta (mert monád*ok*/őket)

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+
+Például fel tudjuk használni ezt a harmadik koordinátát arra, hogy elkülönítsük a pontokat és a vektorokat. Alapból (ránézésre) nem tudnánk megmondani a $(2, 3)$ koordinátájú objektumról, hogy ez egy pontot ábrázol, vagy egy helyvektort, mely a $(2, 3)$ pontba mutat. Viszont most mondhatjuk azt, hogy az összes pont legyen $[x, y, 1]$ alakú, és az összes vektor pedig $[x, y, 0]$.
+
+Más geometriákat is beágyazhatunk ambiens terekbe, például ha nem a $(x, y, 1)$ síkon választanánk pontokat, hanem azt mondanánk, hogy minden pont $x, y$ és $w$ koordinátáinak az $x^2 + y^2 + w^2 = 1$ egyenletet kéne kielégítenie, akkor elliptikus geometriát ábrázolnánk, vagy a hiperbolikus geometria esetén a pontokat az $x^2 + y^2 - w^2 = -1$ egyenlet alapján választanánk. Az a fontos, hogy mind ...
```
minden (x, y, w) pontnak ...

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+
+#### Középpontos vetítés
+
+A vetítési középpontot összekötjük a gömb felületével, és megnézzük, hogy ez az egyenes hol metszi el a síkot amire vetítünk. Ez a metszéspont lesz a gömb felszínén lévő pontnak a vetített pontja. Ha ezt megcsináljuk a gömb minden pontjára, akkor levetítettük a gömböt a síkra. Viszont mivel a vetítési középpont és a gömb felszínének egy pontja által meghatározott egyenes két pontot metszik a gömb ...
```
metszik helyett elég metsz

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+
+#### Sztereografikus vetítés
+
+Az elv teljesen analóg a középpontos vetítéssel, csak itt nem a gömb középpontja lesz a vetítési központ, hanem a gömb egy másik, tetszőleges pontja. Általában ez a gömb valamelyik pólusa szokott lenni, tipikusan a déli. Ekkor a _déli pólus kivételével mindent_ vetítünk, továbbá ez a vetítés _kör-_ és _szögtartó_, de **nem** egyenes- és távolságtartó. Bizonyos értelemben inverze a középpontos vet ...
```
a "pont" redundáns és geometriai kontextusban kicsit zavaró, lehetne "éppen" vagy csak semmi

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+
+## 1.8. Projektív geometria
+
+Ez az egyik legfontosabb geometriai rendszer amit tárgyalunk, mivel az emberi vizuális látás is így értelmezi a világot, szóval amikor grafikai programot csinálunk, akkor ehhez kell igazodni (a GPU mindegyik geometriát támogatja, de projektív geometriában gondolkodik). Az euklideszi geometria tapasztalati alapon készült, viszont a gyakorlatban tapasztaltakat nem írja le pontosan: ha sínekre állun ...
```
grafikai helyett grafikus

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+
+Úgy fogjuk módosítani az euklideszi geometriát, hogy a gyakorlatban tapasztaltakat is pontosan le tudja írni. Mostantól bármilyen két egyenes - legyenek azok párhuzamosak vagy sem - pontosan egy pontban metszik egymást, így a "párhuzamos", mint fogalom értelmét veszti.
+
+Előnye ennek a geometriának - azon felül, hogy pontosabban ábrázolja a valóságban tapasztaltakat - az az, hogy lényegesen leegyszerüsíti a programozást: nem kell olyan edge-caseket vizsgálni, hogy két egyenes párhuzamos, hanem mindig létezik bármely két egyenes metszéspontja. Hátránya viszont az, hogy mivel most már két, euklideszi geometriában párhuzamos egyenes a "végtelenben" metszik egymást,  ...
```
egyből utána "hogy [két egyenes...]" helyett "ahol" szebb

Gilgames32 replied:
"hátránya [...] az, hogy" "az" nem kell megint csak

Gilgames32 replied:
metszik helyett metszi

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+    p \cdot l = 0, \quad l \cdot L^* = 0 \rightarrow l = p \times L^*
+    $$
+
+    ahol $L^*$: $L$-ből a $w$ törlése
```
végére pont

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+    A $p$ ponton átmenő $L$ egyenesre merőleges $l$ egyenes:
+
+    $$
+    p \cdot l = 0, \quad l \cdot L^* = 0 \rightarrow l = p \times L^*
```
valójában a nyelvtanilag helyes úgy lenne, hogy itt a sor végére vesszőt rakunk, mert "ahol"-lal folytatódik

ez egy rakat helyen előjön, illetve az is, hogy a mondat vége a képlettel ér véget; ekkor pontot kéne rakni, viszont szerintem egy std::hallgatónak lehet, hogy ez inkább csak zavaró lenne (nekem is picit), szóval legyen konzisztens, vagy minden képlet végén legyen megfelelő írásjel (ahol mondat része), vagy sehol (ahogy most is van)

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+??? tip Megoldás
+    Távolságot (jelölje $x$) a főkörön mérjük, ott a legrövidebb.
+
+    1\. 45\degree-nál mekkora egy kör kerülete:
```
45° nincs $$-ben, nem jelenik meg jól

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/1.md:
```diff
@@ -0,0 +1,693 @@
+    - [x] $\sinh(3x + 4y + 5) = 0$ egy egyenes
+        - Magyarázat: A $\sinh$ függvény pontosan akkor nulla, ha a bemenete is nulla, tehát csak azok a pontok teljesítik a $\sinh$ féle egyenletet, amik a bemenő egyenes egyenletét is.
+    - [ ] A $3x+4y+5=0$ egyenletű egyenes különbözik a $9x+12y+15=0$ egyenletű egyenestől.  
+        - Magyarázat: Ha $e$ jelöli az első egyenest, akkor a második az $3 \cdot e$ alakú, és ez csak akkor nulla ha $e$ nulla, azaz csak azok a pontok teljesítik a $3 \cdot e = 0$ egyenletet amelyek az $e = 0$ egyenletet is.
```
ha elé vessző

Gilgames32 replied:
amelyek elé vessző

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/2.md:
```diff
@@ -0,0 +1,374 @@
+
+<!-- ez az állítás nem teljes így 1. mert beadhatok neki vec2/3-at és akkor nagyon jól érzem magam, de valójában akkor is vec4 lesz, 2. persze lehet csinálni másmilyen geometriát is, de a GPU ettől még nem fogja úgy számolni nem feltétlenül úgy fogja számolni -->
+
+Ezt követően fix műveleti egységek mennek végbe (vágás, raszterizáció, stb.). Ezek optimalizált hardware műveletek.
```
```suggestion
Ezt követően fix műveleti egységek mennek végbe (vágás, raszterizáció, stb.). Ezek optimalizált hardver műveletek.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/2.md:
```diff
@@ -0,0 +1,374 @@
+Ezt követően fix műveleti egységek mennek végbe (vágás, raszterizáció, stb.). Ezek optimalizált hardware műveletek.
+
+- Vágás: kidobjuk ami "nem látszik".
+- Viewport transzformáció: A normalizált eszközkoordinátákból a képernyő koordináta-rendszerévé ($(0, 0)$ és pl. $(600, 800)$ között) való transzformálás.
```
```suggestion
- Viewport transzformáció: A normalizált eszközkoordinátákból a képernyő koordináta-rendszerévé ($(0, 0)$ és pl. $(800, 600)$ között) való transzformálás.
```
de ez már nagyon kukacoskodás

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/2.md:
```diff
@@ -0,0 +1,374 @@
+- Viewport transzformáció: A normalizált eszközkoordinátákból a képernyő koordináta-rendszerévé ($(0, 0)$ és pl. $(600, 800)$ között) való transzformálás.
+- Raszterizáció: Egy adott sokszög csúcspontjai ismeretében meghatározza az összes olyan pixelt ami benne van a sokszögben, interpolálja a regiszter értékeket (pl. szín).
+
+Ezeket követően már konkrét pixelekkel dolgozik a GPU. A fragment shader a pixelek tényleges színét határozza meg. Végezetül kompozitálás után a pixelek bekerülnek a rasztertárba, ahonnan már egyenesen a képernyőre kerülnek. A kompozitálás azért fontos, hiszen itt dől el, hogy melyik pixelek lennének más pixelek "mögött", az átlátszóság kezeléséhez kiemelten fontos.
```
```suggestion
Ezeket követően már konkrét pixelekkel dolgozik a GPU. A fragment shader a pixelek tényleges színét határozza meg. Végezetül kompozitálás után a pixelek bekerülnek a rasztertárba (frame buffer), ahonnan már egyenesen a képernyőre kerülnek. A kompozitálás azért fontos, mert itt dől el, hogy melyik pixelek lennének más pixelek "mögött", ami az áttetszőség és a takarás kezeléséhez elengedhetetlen.
```

> ahonnan már egyenesen a képernyőre kerülnek

bár ilyen egyszerű lenne xd... de nem baj ne vigyük le szegényeket a compositor rabbitholeon

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/2.md:
```diff
@@ -0,0 +1,374 @@
+- `GL_POINTS`: külön pontokként jeleníti meg a VAO pontjait
+- `GL_LINES`: kettessével összeköti a megadott pontokat, így például az alábbi _szakaszok_ jönnének létre: $(x_0, x_1), (x_2, x_3), ...$
+- `GL_LINE_STRIP`: az adott pontot mindig a következő ponttal köti össze: $(x_0, x_1), (x_1, x_2), (x_2, x_3), ...$
+- `GL_LINE_LOOP`: ugyan az mint a `GL_LINE_STRIP`, csak a legutolsó pontot a legelsővel köti össze
```
```suggestion
- `GL_LINE_LOOP`: ugyan az mint a `GL_LINE_STRIP`, csak a legutolsó pontot a legelsővel is összeköti
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/2.md:
```diff
@@ -0,0 +1,374 @@
+
+## Input kezelése
+
+Az operációs rendszernél a koordináta-rendszer origója a bal felső sarokban található, az $y$ tengely fejjel lefelé van! A `glViewport` viszont a bal alsó sarkot tekinti origónak, tehát valahogy így néz ki a dolog:
```
```suggestion
Az operációs rendszer koordináta-rendszerének origója a bal felső sarokban van, tehát az $y$ tengely **fejjel lefelé** van, "$y$ lefele mutat". A `glViewport` viszont a **bal alsó** sarkot tekinti origónak, és az $y$ tengely "felfele mutat". Valahogy így kell elképzelni:
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/2.md:
```diff
@@ -0,0 +1,374 @@
+    - [x] Az OpenGL nem tud kört rajzolni, mert a művelet felesleges, hiszen a kör közelíthető szabályos sokszöggel.
+        - Magyarázat: Lásd [fentebb](#funkcionális-modell)
+    - [ ] Az OpenGL nem tud kört rajzolni, mert a GPU fejlesztők nem ismerték a kör egyenletét.
+        - Magyarázat: Szerintem a $8$ általánost elvégeszték azért a GPU fejlesztők.
```
xd

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/3.md:
```diff
@@ -0,0 +1,483 @@
+
+#### Lagrange interpoláció
+
+A vezérlőpontokhoz rendelünk csomóértékeket, amik azt fogják jelképezni, hogy melyik "időpillanatban" kell "meglátogatni" az adott pontot. Ha van $n$ pontunk, akkor semmi sem akadályoz meg minket abban, h ezekre illesszünk egy $n + 1000$ fokszámú polinomot, viszont érezhető, hogy nekünk a minimális fokszámú polinom lenne előnyös. Ez az $n-1$-ed fokú lesz.
```
```suggestion
A vezérlőpontokhoz rendelünk csomóértékeket, amik azt fogják jelképezni, hogy melyik "időpillanatban" kell "meglátogatni" az adott pontot. A vezérlőpontokra tetszőlegesen nagy fokszámú polinomot illeszthetünk, viszont egy minimális fokszámú polinom előnyösebb. $n$ vezérlőpont esetén ez egy $n-1$-ed fokú polinom lesz.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/3.md:
```diff
@@ -0,0 +1,483 @@
+    - [x] A Catmull-Rom görbe harmadfokú polinom a kontrollpontok számától függetlenül.
+        - Magyarázat: Igen, mert Hermite-t használ
+    - [ ] A Catmull-Rom görbe bázisfüggvényei nemnegatívak.
+        - Magyarázat: De azok, mert Hermit-et használ
```
```suggestion
        - Magyarázat: De azok, mert Hermite-ot használ.
```

Gilgames32 replied:
mehetne pont az összes végére (idrc, vagy legyen mindenhol vagy sehol)

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/3.md:
```diff
@@ -0,0 +1,483 @@
+    - [x] A Catmull-Rom görbe Hermite interpolációs görbékből épül fel.
+        - Magyarázat: Ez a definíciója.
+    - [x] A Catmull-Rom görbe bázisfüggvények összege 1.
+        - Magyarázat: Igen, mivel a Hermite az általánosítása a Lagrange-nak
```
```suggestion
        - Magyarázat: Igen, mivel a Hermite az általánosítása a Lagrange-nak.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/4.md:
````diff
@@ -0,0 +1,181 @@
+glUniformMatrix4fv(location, 1, GL_FALSE, &matrix[0])
+```
+
+Ha nem egy pontot/szakaszt/háromszöget szeretnénk kirajzolni vagy nem egyenestartó transzformációt akarunk használni, akkor elvégezhetjük a transzformációt, és a végeredményt visszaalakíthatjuk szakaszokra, háromszögekre.
````
```suggestion
Ha nem egy pontot/szakaszt/háromszöget szeretnénk kirajzolni, vagy nem egyenestartó transzformációt akarunk használni, akkor elvégezhetjük a transzformációt, és a végeredményt visszaalakíthatjuk szakaszokra, háromszögekre.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/4.md:
```diff
@@ -0,0 +1,181 @@
+
+## Homogén lineáris transzformációk
+
+Az euklideszi geometriának vannak hiányosságai, például középpontos vetítésre alkalmatlan, ezért projektív geometriát fogunk használni, (möbius féle) homogén koordinátákkal.
```
Möbius szerintem nagy betű de legit nem tudom hogy ilyenkor mi a szabály xd

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+
+Feltételezzük, hogy a virtuális világunk a kétdimenziós Euklideszi sík. Hogyan tudjuk ezt a világot megjeleníteni a képernyőn?
+
+Vannak objektumjaink, amiknek tudjuk a helyét a virtuális világban. Szükségünk lesz egy kamerára, ami azt jelöli majd ki, hogy a virtuális világból mennyit látunk (hol vagyunk, merre nézünk, stb). A végső fénykép pixelekből áll, szóval minden pixelről el kell dönteni, hogy milyen színű legyen. Ha minden objektumhoz rendelünk egy színt, aztán pedig ez az objektum feltűnik egy adott pixelen, akkor  ...
```
```suggestion
Vannak objektumjaink, amiknek tudjuk a helyét a virtuális világban. Szükségünk lesz egy kamerára, ami azt jelöli majd ki, hogy a virtuális világból mennyit látunk (hol vagyunk, merre nézünk stb.). A végső fénykép pixelekből áll, tehát minden pixelről el kell dönteni, hogy milyen színű legyen. Ha minden objektumhoz rendelünk egy színt, és ez az objektum feltűnik egy adott pixelen, akkor tudjuk, hogy annak a pixelnek milyen színűnek kell lennie. (Ha több objektum is verseng egy adott pixelért, azaz van átfedés, akkor explicit prioritást használunk, azaz tartozik egy szám minden objektumhoz, ami megmondja, hogy milyen sorrendben rajzoljuk ki őket.)
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+
+## Pixelvezérelt képszintézis
+
+Induljunk a pixelek felől. A képernyőn véges számú pixel van, tehát elég annyit csinálnunk, hogy minden egyes pixelben eldöntjük, hogy melyik objektum látszódik benne. Transzformáljuk a pixeleket világkoordináta-rendszerbe, és ezután pedig amelyik objektumra landolunk, az van abban a pixelben (figyelembe véve a prioritásukat).
```
```suggestion
Induljunk a pixelek felől. A képernyőn véges számú pixel van, tehát elég annyit csinálnunk, hogy minden egyes pixelben eldöntjük, hogy melyik objektum látszódik benne. Transzformáljuk a pixeleket világkoordináta-rendszerbe, majd amelyik objektumra esik az adott pixel, az fogja meghatározni a színét (figyelembe véve a prioritásukat).
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+Egy sokszög pedig akkor _egyszerű_, ha az oldalai nem keresztezik egymást és "megrajzolható a ceruza felemelése nélkül".
+
+??? abstract Bizonyítás
+    Keressük meg a sokszögünk egy szélső csúcsát, mondjuk az $x$ irányba (tehát a legnagyobb $x$ koordinátájú csúcs). Vegyük a két szomszédját és kössük őket össze. Ekkor két dolog történhet:
```
```suggestion
    Keressük meg a sokszögünk egy szélső csúcsát, mondjuk az $x$ irányban (tehát a legnagyobb $x$ koordinátájú csúcs). Vegyük a két szomszédját és kössük őket össze. Ekkor két dolog történhet:
```

not sure tho

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+
+A vektorizációt követően már csak pontjaink, szakaszaink, és háromszögeink vannak. Ez nekünk azért jó, mert tudunk olyan transzformációkat találni, amik ezeket az alakzatokat "megtartják", azaz háromszöget háromszögbe, szakaszt pedig szakaszba transzformálnak (homogén lineáris transzformációk).
+
+Magukat a transzformációkat elég csak a csúcsokra végrehajtani, nem kell például egy szakasz minden (végtelen sok) pontjára ezt megtenni. Ez nem magától értetődő, direkt azért használunk homogén lineáris transzformációkat, hogy ez a tulajdonság fenn álljon, hiszen míg matematikailag lehet végtelen sok ponttal dolgozni, a GPU-nak ez nehezen menne.
```
```suggestion
Magukat a transzformációkat elég csak a csúcsokra végrehajtani, nem kell például egy szakasz minden (végtelen sok) pontjára ezt megtenni. Ez nem magától értetődő, direkt azért használunk homogén lineáris transzformációkat, hogy ez a tulajdonság fennálljon, hiszen míg matematikailag lehet végtelen sok ponttal dolgozni, a GPU-nak ez nehezen menne.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+
+<!-- NOTE: ide jó lenne több magyarázat az MVP mátrixról -->
+
+Mivel a modell, view és projection transzformációk mind homogén lineáris transzformációk, ezért összevonhatjuk őket mátrixszorzással. Ennek a mátrixnak `MVP` mátrix a neve, és így áll elő:
```
```suggestion
Mivel a modell, view és projection transzformációk mind homogén lineáris transzformációk, ezért összevonhatjuk őket mátrixszorzással. Ez az ún. `MVP` mátrix, és az alábbi módon állítható elő:
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+- `Draw()`: kirajzoljuk az objektumot. Ehhez a már megismert `Camera2D` osztály hívjuk segítségül. A vektorizáció után a modell, view és projection transzformációkat kell elvégezni.
+    - először előállítjuk az `M` mátrixot, mégpedig a korábban tanult alakban: eltolás $\ast$ forgatás $\ast$ skálázás
+    - ezután a kameránkból kinyert `V` és `P` mátrixokkal előállítjuk az `MVP` mátrixot
+    - végezetül pedig feltöltjük a mátrixunkat a GPU-ra mind uniform változót, és meghívjuk a `Geometry` osztály kirajzoló függvényét, ami az OpenGL-es hívásokat ténylegesen megteszi.
```
```suggestion
    - végezetül pedig feltöltjük a mátrixunkat a GPU-ra, mint uniform változót, és meghívjuk a `Geometry` osztály kirajzoló függvényét, ami már az OpenGL függvényeit hívja meg.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
````diff
@@ -0,0 +1,977 @@
+}
+```
+
+Ha ezt mind leprogramoznánk, és írnánk hozzá egy minimális főprogramot (`glApp` leszármazott) ami létrehozza az objektumainkat, akkor tényleg megjelenne a görbe a képernyőn.
````
```suggestion
Ha ezt mind leprogramoznánk, és írnánk hozzá egy minimális főprogramot (`glApp` leszármazott), ami létrehozza az objektumainkat, akkor valóban megjelenne a görbe a képernyőn.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+
+Ha valós idejű animációt szeretnénk, akkor (minimum) harmincszor újra kell rajzolnunk a képernyőt másodpercenként. Egy standard, full HD monitornak $1920 \times 1080 = 2\,073\,600$ pixele van (4K monitoroknak pedig már $8\,294\,400$).
+
+Ha másodpercenként harmincszor akarunk kétmillió pixelt kirajzolni, akkor pixelenként $15$ ns jut átlagosan. Érezzük, hogy itt már sokkal nagyobb teljesítményű algoritmusokat kell majd alkotnunk, olyanokat, amiket szinkron sorrendi hálózatokból meg tudunk alkotni.
```
```suggestion
Ha másodpercenként harmincszor szeretnénk kétmillió pixelt kirajzolni, akkor pixelenként $15$ ns jut átlagosan. Érezzük, hogy itt már sokkal nagyobb teljesítményű algoritmusokat kell majd alkotnunk, olyanokat, amiket szinkron sorrendi hálózatokból meg tudunk alkotni.
```

"meg tudunk alkotni" kicsit ehh de nem jut eszembe jobb most

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+
+Itt még mindig kerekíteni kell, de megspórulunk sok felesleges szorzást.
+
+A kerekítés gyorsítása érdekében sok hardware megoldás született, az egyik ilyen a fixpontos inkrementális számítási elv:
```
```suggestion
A kerekítés gyorsítása érdekében sok hardver megoldás született, az egyik ilyen a fixpontos inkrementális számítási elv:
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+
+A fő felismerésünk az, hogy ha egy tulajdonság igaz egy pixelre, akkor nagy eséllyel a szomszédjaira is, hacsak valamilyen határt át nem lépünk. Csoportosítsuk a pixeleinkát _pásztákba_, azaz hosszú vízszintes sorokba. Ekkor egy pásztán belül az elején sok pixel nem lesz benne a háromszögben, aztán lesz egy határpont, ahol átbillenünk, és hirtelen minden pixel a háromszögön belül van, aztán meg e ...
+
+Nézzük a konkrét algoritmust: messük el a háromszögünket vízszintes pásztákkal, és számoljuk ki a két metszéspont $x$ koordinátáit (az $y$ ugyan az lenne, hiszen vízszintes az egyenesünk). Ekkor csak eközött a két $x$ koordináta között kell kiszínezni a pixeleket, nem kell semmi extra műveletet elvégezni a pásztára, ha ez a két érték megvan már.
```
a messük egy létező szó, im flabbergasted
```suggestion
Nézzük a konkrét algoritmust: messük el a háromszögünket vízszintes pásztákkal, és számoljuk ki a két metszéspont $x$ koordinátáit (az $y$ ugyan az lenne, hiszen az egyenesünk vízszintes). Ekkor csak eközött a két $x$ koordináta között kell kiszínezni a pixeleket, nem kell semmi extra műveletet elvégezni a pásztára, ha ez a két érték megvan már.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/5.md:
```diff
@@ -0,0 +1,977 @@
+    ![quiz](../img/chapter_5/5_kviz_csillag.png){width=100%}
+
+??? tip Megoldás
+    Mivel nincs diagonálja, ezért füle sincs (mert a fül az amikor egy pont szomszédai összeköthetők diagonállal).
```
```suggestion
    Mivel nincs diagonálja, ezért füle sincs (mert a fül az, amikor egy pont szomszédai összeköthetők diagonállal).
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/12.md:
````diff
@@ -0,0 +1,1002 @@
+}
+```
+
+Részletesebb magyarázathoz ajánlom az [előadásvideó ezen részét](https://youtu.be/4-5_UzT6SFQ?si=9OIrHGFQrlBpCLUV&t=2317).
````
```suggestion
Részletesebb magyarázathoz ajánlom az [előadásvideó ezen részét](https://youtu.be/4-5_UzT6SFQ&t=2317).
```
EVIL FUCKING TRACKER

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/12.md:
````diff
@@ -0,0 +1,1002 @@
+}
+```
+
+A fenti kód részletesebb magyarázatát, illetve a hozzá tartozó GPU implementációt az erről szóló [Grafika videóban](https://youtu.be/4-5_UzT6SFQ?si=y_a7fh_TWTc9ZbYI&t=3150) találod.
````
```suggestion
A fenti kód részletesebb magyarázatát, illetve a hozzá tartozó GPU implementációt az erről szóló [Grafika videóban](https://youtu.be/4-5_UzT6SFQ&t=3150) találod.
```
EVIL FUCKING TRACKER

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/6.md:
```diff
@@ -0,0 +1,575 @@
+
+A textúrázáshoz használt kép elemei a _texelek_ (mint ahogy a pixel a "picture element"-ből származik, a texel a "texture element"-ből).
+
+A textúrázást elég csak háromszögekre megvalósítani, hiszen az összes többi bonyolultabb felületet háromszögekből építjük fel. Az a célunk, hogy a textúraterünk háromszögeit valahogy megfeleltessük a modellterünk háromszögeivel. Az egyik ilyen megfeleltetés, amit a leggyakrabban használunk az a _lineáris leképzés_.
```
```suggestion
A textúrázást elég csak háromszögekre megvalósítani, hiszen az összes többi bonyolultabb felületet háromszögekből építjük fel. Az a célunk, hogy a textúraterünk háromszögeit valahogy megfeleltessük a modellterünk háromszögeivel. A leggyakrabban használt egyik ilyen megfeleltetés a _lineáris leképzés_.
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/6.md:
```diff
@@ -0,0 +1,575 @@
+
+## Lineáris leképzés
+
+A fő ötlet az, hogy _"bármely két háromszög között létezik olyan affin transzformáció, ami az egyik háromszöget a másikba viszi át"_ és ez az affin transzformáció Descartes koordinátákban kifejezve egy _lineáris kifejezés_, azaz mind $u$, mind $v$ elsőfokú hatványokként jelennek meg. Más szóval az a megfeleltetést keressük, ami az $(u,v)$ textúra térbeli koordinátákat megfelelteti a modell térbel ...
```
```suggestion
Bármely két háromszög között létezik olyan affin transzformáció, ami az egyik háromszöget a másikba viszi át. Ez az affin transzformáció Descartes koordinátákban kifejezve egy _lineáris kifejezés_, azaz $u$ és $v$ elsőfokú hatványokként szerepelnek. Tehát azt a megfeleltetést keressük, ami az $(u,v)$ textúra térbeli koordinátákat megfelelteti a modell térbeli $(x,y,z)$ koordinátáknak, és a kapcsolatuk lineáris, azaz:
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/6.md:
```diff
@@ -0,0 +1,575 @@
+
+## Textúra szűrés
+
+Eljutottunk odáig, hogy minden egyes pixelt hatékonyan át tudunk transzformálni a textúra térbe. Viszont [nem triviális](https://cdn.discordapp.com/emojis/1384476146636161065.webp?size=240), hogy ez alapján a transzformált pont alapján melyik texelt válasszuk ki. Ez a kiválasztási folyamat a _textúra szűrés_. E során a pixelek középpontjait transzformáljuk a textúra térbe, és megnézzük, hogy mely ...
```
```suggestion
Eljutottunk odáig, hogy minden egyes pixelt hatékonyan át tudunk transzformálni a textúra térbe. Viszont [nem triviális](https://cdn.discordapp.com/emojis/1384476146636161065.webp?size=240), hogy ez alapján a transzformált pont alapján melyik texelt választjuk ki. Ez a kiválasztási folyamat a _textúra szűrés_. Ennek során a pixelek középpontjait transzformáljuk a textúra térbe, és megnézzük, hogy melyik texel középpontjához vannak a legközelebb. Ez a `GL_NEAREST` (nearest neighbor), más néven _matematikai mintavételezés_. Ez alapesetben jól működik, viszont két olyan eset is felmerülhet, amikor nekünk ez nem megfelelő:
```

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/7.md:
```diff
@@ -0,0 +1,555 @@
+A való világban egy fényforrás egy véges felülettel rendelkezik, azaz egy kiterjed test. Viszont egy ilyen fényforrás minden pontjából bocsájtana ki fotonokat, tehát ha egy felület egy pontján az összes oda beeső fénysugárral számolni akarnánk, akkor integrálnunk kéne a fényforrás felülete mentén.
+
+!!! quote Idézet
+    $\text{"Integrálni pedig általában nem szeretünk [...]"}$
```
banger xd

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/7.md:
```diff
@@ -0,0 +1,555 @@
+f_r(\bm{L}, r, \bm{V}) \stackrel{\text{def}}{=} \frac{L(r, \bm{V})}{L^{\text{in}}(r, \bm{L}) \cdot \cos(\theta^{\text{in}})}
+$$
+
+akkor sok kellemes tulajdonságát tudjuk élvezni, az egyik például a Helmholtz törvény:
```
> Egyik előnye, hogy így szét bontjuk a geometriai és anyagi tényezőket, hogy ha a BRDF-et így definiáljuk:

[képlet] + ez a szöveg nem rak ki egy értelmes mondatot ezt az egészet az előzővel együtt át kéne fogalmazni

Gilgames32 commented on docs/notes/sem4/computer_graphics/lectures/7.md:
```diff
@@ -0,0 +1,555 @@
+Ezt a spekuláris huplit hozzáadva a korábbi diffúzhoz kapunk egy úgy referencia sugársűrűséget. Ez a képlet már nem szimmetrikus (sérti a fizikát):
+
+$$
+\boxed{
```
kell amúgy a zárójel cos+theta^in-re? mmint korábban is volt és ott is volt zárójel de kinda felesleges mert ez egyetlen tag, és pl a cos+rho nincs zárójelben

Gilgames32 commented:
chat ezzel idk mi legyen a gh retardált és nem engedi leadni a reviewt de like meddig várjunk vele? adtam le ticketet de még semmi

levyry commented on docs/notes/sem4/computer_graphics/lectures/7.md:
```diff
@@ -0,0 +1,555 @@
+# 3D képszintézis
```
test comment

levyry reviewed (commented):
test review

levyry commented:
grrrr >:(

levyry commented:
Since GitHub won't let @Gilgames32 post his review of the first seven chapters (billion dollar company btw) the review will have to stay private. Don't worry, he got it to me through... [other methods...](https://cdn.discordapp.com/attachments/1354827475615613161/1509526090425696326/image.png?ex=6a197f39&is=6a182db9&hm=07a974012c6154b675e59b0ad5454c8986ebc5659914307fd8fbd84230326c56&animated=true) anyways, I will address the review in the following days, most likely during this weekend.

Everyone say "Thank you Kapucni!"

Kristof-me reviewed (dismissed):
Test message with approval

levyry committed:
implement requested changes

I also rewrote some parts of the math appendix.

Co-authored-by: Kovács Bálint <balintkovacs11@gmail.com>

Gilgames32 reviewed (dismissed):
never using github ever again
